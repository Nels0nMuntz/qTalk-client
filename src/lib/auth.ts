import { NextAuthOptions, getServerSession } from 'next-auth';
import GoogleProvider from 'next-auth/providers/google';
import CredentialsProvider from 'next-auth/providers/credentials';
import { PrismaAdapter } from '@auth/prisma-adapter';
import { nanoid } from 'nanoid';
import { compare } from 'bcrypt';
import { db } from './db';

export const authoOptions: NextAuthOptions = {
  adapter: PrismaAdapter(db),
  session: {
    strategy: 'jwt',
  },
  pages: {
    signIn: '/sign-in',
  },
  secret: process.env.NEXTAUTH_SECRET,
  // debug: process.env.NODE_ENV === 'development',
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID!,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
    }),
    CredentialsProvider({
      type: 'credentials',
      name: 'credentials',
      credentials: {
        email: { label: 'Email', type: 'text' },
        password: { label: 'Password', type: 'text' },
      },
      async authorize(credentials) {
        if (!credentials?.email || !credentials.password) {
          return null;
        }
        const userbyEmail = await db.user.findFirst({
          where: {
            email: credentials.email,
          },
        });

        if (!userbyEmail || !userbyEmail.password) {
          return null;
        }

        const isPasswordValid = await compare(
          credentials.password,
          userbyEmail.password,
        );

        if (!isPasswordValid) {
          return null;
        }

        return userbyEmail;
      },
    }),
  ],
  callbacks: {
    async session({ token, session }) {
      if (token) {
        session.user.id = token.id;
        session.user.name = token.name;
        session.user.email = token.email;
        session.user.image = token.picture;
        session.user.username = token.username;
      }

      return session;
    },

    async jwt({ token, user }) {
      const dbUser = await db.user.findFirst({
        where: {
          email: token.email,
        },
      });

      if (!dbUser) {
        token.id = user!.id;
        return token;
      }

      if (!dbUser.username) {
        await db.user.update({
          where: {
            id: dbUser.id,
          },
          data: {
            username: nanoid(10),
          },
        });
      }

      return {
        id: dbUser.id,
        name: dbUser.name,
        email: dbUser.email,
        picture: dbUser.image,
        username: dbUser.username,
      };
    },
  },
};

export const getAuthSession = () => getServerSession(authoOptions);
