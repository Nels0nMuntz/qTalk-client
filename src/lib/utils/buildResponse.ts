export const buildResponse = (body: string | object, init?: ResponseInit) => {
  const responseBody =
    typeof body === 'string'
      ? JSON.stringify({ message: body })
      : JSON.stringify(body);
  return new Response(responseBody, init);
};
