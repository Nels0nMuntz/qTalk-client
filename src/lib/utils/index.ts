export * from './buildResponse';
export * from './utils';
export * from './notify';
