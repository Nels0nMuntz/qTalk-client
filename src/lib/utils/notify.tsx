import toast, { ToastOptions, Toast } from 'react-hot-toast';
import { CheckCircle, XCircle, X } from 'lucide-react';
import React from 'react';

type NotifyVariants = 'success' | 'error';

const icons: Record<NotifyVariants, JSX.Element> = {
  success: <CheckCircle color="#a3e635" />,
  error: <XCircle color="#f87171" />,
};

export const notify = ({
  title,
  description,
  render,
  variant,
  options,
}: {
  title?: string;
  description?: string;
  render?: (ctx: Toast) => React.ReactNode;
  variant: 'success' | 'error';
  options?: ToastOptions;
}) => {
  const Icon = icons[variant];
  return toast(
    (t) => (
      <div className="flex items-start">
        <div className="shrink-0 mr-4">{Icon}</div>
        <div className="flex flex-col gap-y-1">
          <span className="text-sm font-semibold">{title}</span>
          <span className="text-sm" onClick={() => toast.dismiss(t.id)}>
            {description}
          </span>
          {render && render(t)}
        </div>
        <div className="ml-4 shrink-0">
          <button
            className="text-zinc-500 hover:text-zinc-700 transition-all rounded-md focus:outline focus:outline-offset-1 focus:outline-2 focus:outline-indigo-500"
            type="button"
            onClick={() => toast.dismiss(t.id)}
          >
            <span className="sr-only">Close</span>
            <X size={20} />
          </button>
        </div>
      </div>
    ),
    options,
  );
};
