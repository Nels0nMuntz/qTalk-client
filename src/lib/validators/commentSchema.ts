import { z } from 'zod';

export const commentSchema = z.object({
  postId: z.string(),
  replyToId: z.string().optional(),
  text: z.string(),
});

export type CreateCommentPayload = z.infer<typeof commentSchema>;
