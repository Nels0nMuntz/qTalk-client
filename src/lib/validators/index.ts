export * from './signInFormSchema';
export * from './signUpFormSchema';
export * from './subtalkSchema';
