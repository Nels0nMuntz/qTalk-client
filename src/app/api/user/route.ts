import { db } from '@/lib/db';
import { buildResponse } from '@/lib/utils';
import { signUpFormSchema } from '@/lib/validators';
import { hash } from 'bcrypt';
import { z } from 'zod';

export async function POST(req: Request) {
  try {
    const body = await req.json();
    const { email, username, password } = signUpFormSchema.parse(body);

    const userByEmail = await db.user.findFirst({
      where: {
        email,
      },
    });
    const userByUsername = await db.user.findFirst({
      where: {
        username,
      },
    });
    if (userByEmail || userByUsername) {
      return buildResponse('Email or username are taken', { status: 422 });
    }

    const hashedPassword = await hash(password, 10);
    await db.user.create({
      data: {
        email,
        username,
        password: hashedPassword,
      },
    });
    return buildResponse('User was created', { status: 201 });
  } catch (error) {
    console.log(error);

    if (error instanceof z.ZodError) {
      return new Response(error.message, { status: 422 });
    }

    return buildResponse('Internal server error', { status: 500 });
  }
}
