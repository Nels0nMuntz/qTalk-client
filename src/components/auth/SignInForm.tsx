'use client';

import React from 'react';
import Link from 'next/link';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { Button } from '../ui/button';
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from '../ui/form';
import { Input } from '../ui/input';
import { SignInFormSchema, signInFormSchema } from '@/lib/validators';
import { Checkbox } from '../ui/checkbox';
import { signIn } from 'next-auth/react';
import { useRouter, useSearchParams } from 'next/navigation';
import { notify } from '@/lib/utils';

export default function SignInForm() {
  const router = useRouter();
  const searchParams = useSearchParams();
  const [isSubmitting, setIsSubmitting] = React.useState(false);
  const form = useForm<SignInFormSchema>({
    resolver: zodResolver(signInFormSchema),
    defaultValues: {
      email: '',
      password: '',
      rememberMe: false,
    },
  });

  const onSubmit = async (values: SignInFormSchema) => {
    setIsSubmitting(true);

    try {
      const response = await signIn('credentials', {
        ...values,
        redirect: false,
        callbackUrl: searchParams.get('callbackUrl') || '/',
      });

      if (response?.ok) {
        router.push('/');
        router.refresh();
      } else {
        notify({
          title: 'Email or password is invalid',
          variant: 'error',
        });
      }
    } catch (error) {
      notify({
        title: 'There was a problem',
        description: 'There was an error loggong in',
        variant: 'error',
      });
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <Form {...form}>
      <form
        onSubmit={form.handleSubmit(onSubmit)}
        className="space-y-4 text-left"
      >
        <FormField
          control={form.control}
          name="email"
          render={({ field }) => (
            <FormItem className="space-y-1">
              <FormLabel>Email</FormLabel>
              <FormControl>
                <Input placeholder="your@email.com" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="password"
          render={({ field }) => (
            <FormItem className="space-y-1">
              <FormLabel>Password</FormLabel>
              <FormControl>
                <Input placeholder="at least 6 characters" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="rememberMe"
          render={({ field }) => (
            <FormItem className="flex items-center gap-2 space-y-0">
              <FormControl>
                <Checkbox
                  {...field}
                  value={field.value?.toString()}
                  checked={field.value}
                  onCheckedChange={field.onChange}
                />
              </FormControl>
              <FormLabel className="mt-0">Remember Me</FormLabel>
              <FormMessage />
            </FormItem>
          )}
        />
        {/* <Link
          href="/forgot-password"
          className="block text-sm font-medium text-center text-orange-600"
        >
          Forgot Password ?
        </Link> */}
        <Button
          className="w-full text-white bg-rose-400 rounded-lg hover:bg-rose-500 transition-colors"
          isLoading={isSubmitting}
        >
          Sign In
        </Button>
      </form>
    </Form>
  );
}
