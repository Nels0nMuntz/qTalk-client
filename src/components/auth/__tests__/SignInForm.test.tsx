import { render, screen, waitFor } from '@testing-library/react';
import user from '@testing-library/user-event';
import { signIn } from 'next-auth/react';
import { useRouter, useSearchParams } from 'next/navigation';
import { notify } from '@/lib/utils';
import SignInForm from '../SignInForm';
import { SignInFormSchema } from '@/lib/validators';


jest.mock('next-auth/react', () => ({
  signIn: jest.fn(),
}));

const mockPush = jest.fn()
const mockRefresh = jest.fn()
jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: mockPush,
    refresh: mockRefresh,
  }),
  useSearchParams: () => ({
    get: jest.fn(() => '/'),
  }),
}));

jest.mock('../../../lib/utils/notify.tsx', () => ({
  notify: jest.fn(),
}));

class ResizeObserver {
  observe() {}
  unobserve() {}
  disconnect() {}
}

const queryEmailHelperText = (container: HTMLElement) => {
  return container.querySelector('form > div:first-child > input + p');
};

const queryPasswordHelperText = (container: HTMLElement) => {
  return container.querySelector('form > div:nth-child(2) > input + p');
};

const submitForm = async (formData: SignInFormSchema) => {

  render(<SignInForm />);

  const emailInput = await screen.findByLabelText('Email');
  const passwordInput = await screen.findByLabelText('Password');
  const checkbox = await screen.findByRole('checkbox', {
    name: 'Remember Me',
  });
  const submitButton = await screen.findByRole('button', {
    name: /sign in/i,
  });

  await user.type(emailInput, formData.email);
  await user.type(passwordInput, formData.password);
  await user.click(checkbox); 
  await user.click(submitButton);
}

describe('SignInForm component', () => {
  beforeEach(() => {
    jest.clearAllMocks(); // Clear mock calls before each test
  });
  window.ResizeObserver = ResizeObserver;
  it('renders correctly', async () => {
    const { container } = render(<SignInForm />);

    const form = await screen.findByRole('form', { name: 'signin form' });

    expect(form).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
  it('renders emain validation errors correctly', async () => {
    const { container } = render(<SignInForm />);

    const emailInput = await screen.findByLabelText('Email');
    const submitButton = await screen.findByRole('button', {
      name: /sign in/i,
    });

    expect(emailInput).toHaveValue('');
    await user.click(submitButton);
    expect(queryEmailHelperText(container)).toHaveTextContent('Invalid email');

    await user.type(emailInput, 'wrong@email');
    expect(queryEmailHelperText(container)).toHaveTextContent('Invalid email');

    await user.clear(emailInput);
    await user.type(emailInput, 'right@gmail.com');
    expect(queryEmailHelperText(container)).toBeFalsy();
  });

  it('renders password validation errors correctly', async () => {
    const { container } = render(<SignInForm />);

    const passwordInput = await screen.findByLabelText('Password');
    const submitButton = await screen.findByRole('button', {
      name: /sign in/i,
    });

    expect(passwordInput).toHaveValue('');
    await user.click(submitButton);
    expect(queryPasswordHelperText(container)).toHaveTextContent(
      'String must contain at least 6 character(s)',
    );

    await user.type(passwordInput, 'wrong');
    expect(queryPasswordHelperText(container)).toHaveTextContent(
      'String must contain at least 6 character(s)',
    );

    await user.clear(passwordInput);
    await user.type(passwordInput, 'toolongpasswordtoolongpassword');
    expect(queryPasswordHelperText(container)).toHaveTextContent(
      'String must contain at most 20 character(s)',
    );

    await user.clear(passwordInput);
    await user.type(passwordInput, 'goodpassword');
    expect(queryPasswordHelperText(container)).toBeFalsy();
  });

  it('toggles checkbox value', async () => {
    render(<SignInForm />);

    const checkbox = await screen.findByRole('checkbox');

    expect(checkbox).toHaveValue('false');

    await user.click(checkbox);

    expect(checkbox).toHaveValue('true');

    await user.click(checkbox);

    expect(checkbox).toHaveValue('false');
  });

  it('submits form with form values', async () => {
    const formData = {
      email: 'jenna@jenna.com',
      password: 'jennajenna',
      rememberMe: true,
    };

    await submitForm(formData);

    expect(signIn).toHaveBeenCalledWith('credentials', {
      ...formData,
      redirect: false,
      callbackUrl: '/',
    });
  });

  it('redirects to home after successful signin', async () => {
    const signInMock = signIn as jest.MockedFunction<typeof signIn>;
    signInMock.mockResolvedValue({
      ok: true,
      error: null,
      status: 200,
      url: '/'
    });

    const formData = {
      email: 'jenna@jenna.com',
      password: 'jennajenna',
      rememberMe: true,
    };

    await submitForm(formData);

    expect(mockPush).toHaveBeenCalledWith('/');
    expect(mockRefresh).toHaveBeenCalled();
    expect(notify).not.toHaveBeenCalled();
  });

  it('show error notification if signin failed', async () => {
    const signInMock = signIn as jest.MockedFunction<typeof signIn>;
    signInMock.mockResolvedValue({
      ok: false,
      error: null,
      status: 400,
      url: '/'
    });

    const formData = {
      email: 'jenna@jenna.com',
      password: 'jennajenna',
      rememberMe: true,
    };

    await submitForm(formData);

    expect(mockPush).not.toHaveBeenCalled();
    expect(mockRefresh).not.toHaveBeenCalled();
    expect(notify).toHaveBeenCalled();
  })
});
