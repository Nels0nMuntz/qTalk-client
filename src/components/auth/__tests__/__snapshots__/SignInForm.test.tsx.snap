// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm component renders correctly 1`] = `
<div>
  <form
    aria-label="signin form"
    class="space-y-4 text-left"
  >
    <div
      class="space-y-1"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Email
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="email"
        placeholder="your@email.com"
        value=""
      />
    </div>
    <div
      class="space-y-1"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Password
      </label>
      <input
        aria-describedby=":r1:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r1:-form-item"
        name="password"
        placeholder="at least 6 characters"
        value=""
      />
    </div>
    <div
      class="flex items-center gap-2 space-y-0"
    >
      <button
        aria-checked="false"
        aria-describedby=":r2:-form-item-description"
        aria-invalid="false"
        class="peer h-4 w-4 shrink-0 rounded-sm border border-primary ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground"
        data-state="unchecked"
        id=":r2:-form-item"
        role="checkbox"
        type="button"
        value="false"
      />
      <input
        aria-hidden="true"
        name="rememberMe"
        style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
        tabindex="-1"
        type="checkbox"
        value="false"
      />
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 mt-0"
        for=":r2:-form-item"
      >
        Remember Me
      </label>
    </div>
    <button
      class="inline-flex items-center justify-center text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 h-10 px-4 py-2 w-full text-white bg-rose-400 rounded-lg hover:bg-rose-500 transition-colors"
    >
      Sign In
    </button>
  </form>
</div>
`;
